File: python/uxandeh/hash.py
"""Hash Module.

This module implements the main hashing structure of the African Fractal Hash
(uxandeh) using Merkle-Damgård padding for 32-byte block sizes.

All code and comments in this module follow British English spelling conventions.
"""

from typing import Tuple
from .compression import DEFAULT_IV, PRIME_P, compress_block

def pad(data: bytes) -> bytes:
    """Pad the input data using Merkle-Damgård padding tailored for 32-byte blocks.

    Appends a 0x80 byte, then a variable number of 0x00 bytes, and finally
    the length of the original message in bits as a 64-bit big-endian integer.
    The total length of the padded message will be a multiple of 32 bytes.
    """
    original_len_bits = len(data) * 8
    
    # Append the 0x80 byte
    padded = bytearray(data)
    padded.append(0x80)
    
    # Calculate padding bytes needed
    # (len(data) + 1 + K) % 32 == 24
    current_mod = len(padded) % 32
    if current_mod <= 24:
        k = 24 - current_mod
    else:
        k = 32 - (current_mod - 24)
        
    padded.extend(b'\x00' * k)
    
    # Append the length in bits as an 8-byte big-endian integer
    padded.extend(original_len_bits.to_bytes(8, "big"))
    
    return bytes(padded)


def uxandeh(data: bytes) -> bytes:
    """Compute the 256-bit African Fractal Hash (uxandeh) digest of the input bytes.

    Args:
        data: The input message as bytes.

    Returns:
        A 32-byte (256-bit) hash digest.
    """
    # 1. Pad the message
    padded_msg = pad(data)
    
    # 2. Initialise the hash state
    state = DEFAULT_IV
    
    # 3. Process each 32-byte block
    for offset in range(0, len(padded_msg), 32):
        block = padded_msg[offset : offset + 32]
        state = compress_block(state, block)
        
    # 4. Finalise digest: convert state (four 64-bit integers) to 32 bytes
    digest = bytearray()
    for val in state:
        # Each value is modulo PRIME_P, so it fits in a 64-bit integer
        digest.extend(val.to_bytes(8, "big"))
        
    return bytes(digest)

File: rust/src/hash.rs
/// Hash module.
///
/// Implements Merkle-Damgård padding and the compression processing loop.
///
/// All code and comments follow British English spelling conventions.

use crate::compression::{compress_block, DEFAULT_IV};

pub fn pad(data: &[u8]) -> Vec<u8> {
    let original_len_bits = (data.len() as u64) * 8;
    
    let mut padded = Vec::with_capacity(data.len() + 32);
    padded.extend_from_slice(data);
    padded.push(0x80);

    let current_mod = padded.len() % 32;
    let k = if current_mod <= 24 {
        24 - current_mod
    } else {
        32 - (current_mod - 24)
    };

    padded.resize(padded.len() + k, 0x00);
    padded.extend_from_slice(&original_len_bits.to_be_bytes());
    padded
}

pub fn uxandeh(data: &[u8]) -> [u8; 32] {
    let padded = pad(data);
    let mut state = DEFAULT_IV;

    for chunk in padded.chunks_exact(32) {
        let block: &[u8; 32] = chunk.try_into().unwrap();
        state = compress_block(state, block);
    }

    let mut digest = [0u8; 32];
    digest[0..8].copy_from_slice(&state.0.to_be_bytes());
    digest[8..16].copy_from_slice(&state.1.to_be_bytes());
    digest[16..24].copy_from_slice(&state.2.to_be_bytes());
    digest[24..32].copy_from_slice(&state.3.to_be_bytes());
    digest
}
